a10networks/a10-octavia
# [Stack 1427, 1452, 1453][BUG][STACK-1451]: Refactor pool flow for good reverts during Pool Create, Delete and Update

## Description
- In order to resolve the issue of locking into `PENDING_*` state of flows, once error is encountered, it required us to add proper `raise Exceptions` on VThunder tasks, to enable good revert calls.
- Added and tested proper raise for Pool create, delete and update flows.

Bug Fix:
- Severity Level : **Critical**
- During update operation of Pool, while setting `--session-persistence` object, we encounter `AttributeError`, which say `pool.session_persistence` is of `dict` type.

## Jira Ticket
https://a10networks.atlassian.net/browse/STACK-1427
https://a10networks.atlassian.net/browse/STACK-1452
https://a10networks.atlassian.net/browse/STACK-1453
BUG
https://a10networks.atlassian.net/browse/STACK-1451

## Technical Approach
- Since it is expected that axapi call will throw `ACOSException` and `ConnectionError` possibly, catching them only, in create*, update* and delete* tasks.
- Added proper try-except and LOGs for every axapi calls.
- In `PoolUpdate` task, resolved the session persistence bug which was occuring because of converting `pool` object to dictionary using `pool.__dict__.update(update_dict)`. Resolved that.
- Removed unused code from pool flows.


## Limitations
- **NO reverts are added for UPDATE and DELETE tasks** 
- If reverts are occurring during UPDATE commands, the vthunder config will get updated for all successful vthunder tasks executed before a failed task, as there is no revert task written
- @hthompson-a10 Let me know your thoughts on this behaviour for `PoolDelete` flow:
If pool (having children) is deleted because of an axapi error in pool delete task, then following things happen:
   a) LB in db --> `ACTIVE` state
   b) Listener in db --> `ACTIVE` state 
   c)Pool  in db --> 'ERROR' state 
   **d) Member in db --> `ACTIVE` state
   e) Health_monitor in db --> `ACTIVE` state
   f) Pool on vThunder ---> remains
   g) Member on vThunder --> gets removed
   h) Health_monitor on vThunder ---> gets removed**


## Config Changes
None



## Test Cases
- Create Pool, if any task raises an error, the entire flow reverts and `pool` goes into `ERROR` state. The 'loadbalancer' remains in `ACTIVE` state. No new config is seen on vthunder.
- Update Pool, if any task raises an error, the entire flow reverts and `pool` goes into `ERROR` state. The 'loadbalancer' remains in `ACTIVE` state. **_The config on vthunder is updated for all successful vthunder related tasks executed before the failed task._**
-  Delete Pool, if any task raises an error, the entire flow reverts and `pool` goes into `ERROR` state. The 'loadbalancer' remains in `ACTIVE` state. **_The config on vthunder is updated for all successful vthunder related tasks executed before the failed task._**



## Manual Testing
_**Create Pool**_
- Add `raise ConnectionError` in `PoolCreate` task or `HandleSessionPersistenceDelta` and create a simple pool
eg : `openstack loadbalancer pool create --protocol HTTP --lb-algorithm ROUND_ROBIN --listener l1 --name pool1`

-----> _**Check following**_ 
a) `openstack loadbalancer pool list` -> should go in `ERROR` state
b) `openstack loadbalancer list` --> should remain in `ACTIVE` state
c)  On vThunder cli , no `service-group` config is present.

- Add `raise ConnectionError` in `PoolCreate` task or `HandleSessionPersistenceDelta` task and create a simple pool with Session Persistence
eg: `openstack loadbalancer pool create --protocol HTTP --listener listener1 --session-persistence type=HTTP_COOKIE --lb-algorithm SOURCE_IP --name pool1`

-----> _**Check following**_ 
a) `openstack loadbalancer pool list` -> should go in `ERROR` state
b) `openstack loadbalancer list` --> should remain in `ACTIVE` state
c)  On vThunder cli , **no** `service-group` config is present.
d)  On vThunder cli , **no** `slb template persist cookie 551d9****` config is present.


**_Update Pool_**
- Add `raise ConnectionError` in `PoolUpdate` task or `HandleSessionPersistenceDelta` task and update a pool
eg: `openstack loadbalancer pool set --lb-algorithm SOURCE_IP pool1`

-----> _**Check following**_ 
a) `openstack loadbalancer pool list` -> should go in `ERROR` state
b) `openstack loadbalancer list` --> should remain in `ACTIVE` state
c)  On vThunder cli , the config will get updated with all the successful vthunder tasks,  that were present before the FAILED task. **They wont be reverted. (Since no reverts are present for update tasks)**

- Add `raise ConnectionError` in `PoolUpdate` task or `HandleSessionPersistenceDelta`  task and update the pool with new Session Persistence
eg: `openstack loadbalancer pool set --lb-algorithm SOURCE_IP --session-persistence type=APP_COOKIE,cookie_name=test_cookie pool1`

**_Delete Pool_**
- Add `raise ConnectionError` in `PoolDelete` task or `DeleteSessionPersistence` task and delete a pool
eg: `openstack loadbalancer pool delete pool1`

-----> _**Check following**_ 
a) `openstack loadbalancer pool list` -> should go in `ERROR` state
b) `openstack loadbalancer list` --> should remain in `ACTIVE` state
c)  On vThunder cli , the config will get updated with all the successful vthunder tasks,  that were present before the FAILED task. **They wont be reverted. (Since no reverts are present for delete tasks)**

_**Please check in Limitations section for behaviour of pool children during reverts.**_

## Screenshots 
Example snippet for Logs in case of errors

In case of execute failure
![service_group_revert](https://user-images.githubusercontent.com/52992745/87798317-79466800-c869-11ea-9472-f4b83897e472.PNG)

In case of revert failure
![image](https://user-images.githubusercontent.com/52992745/87797946-063cf180-c869-11ea-99f8-34a08e8aad5c.png)
